Enable `Allreduce`/`Broadcast` TF ops by Intel® oneAPI (DPC++ compiler/oneCCL/IMPI included) software stack on Intel GPU device.

LuFinch opened the pull request:
## Motivation

This PR aims to bring **Intel oneAPI (DPC++ compiler/oneCCL/IMPI included)** software stack into Horovod to make distributed TensorFlow/PyTorch workloads run on [Intel GPU devices](https://www.intel.com/content/www/us/en/products/details/discrete-gpus/data-center-gpu.html) together with [Intel® Extension for TensorFlow*](https://github.com/intel/intel-extension-for-tensorflow) and  [Intel® Extension for PyTorch*](https://github.com/intel/intel-extension-for-pytorch), which are open-source solutions to run TensorFlow and PyTorch applications on Intel AI hardware.

**Intel oneAPI** is an initiative to deliver a unified and open programming experience to developers on any architecture, such as CPUs, GPUs, FPGAs, and AI accelerators. Intel is promoting oneAPI as the industry standard and delivers Intel oneAPI products (an implementation of oneAPI standard) for high-performance heterogeneous computing. 

DPC++ compiler/oneCCL/IMPI packaged into oneAPI Base Tookit are core set of tools and libraries for developing high-performance computing and cross-devices communication applications. **DPC++** compiler is an extension of C++ for heterogeneous computing and based on SYCL. **OneCCL** (oneAPI Collective Communications Library) and **IMPI** provide high-performance communication patterns for distributed workloads in Intel® GPU cluster. 

To bring Intel GPU devices into TensorFlow and PyTorch community for AI workload acceleration, Intel promoted a new user visible **`XPU` device type** as a device abstraction for Intel heterogeneous computation architectures and implemented corresponding device runtime in **Intel® Extension for TensorFlow*** and **Intel® Extension for PyTorch***. 

This diagram provides a summary of the above package ecosystem. 
<img src="https://github.com/intel-innersource/frameworks.ai.horovod/assets/20787576/22bf1207-c04c-4a16-8c48-283ddf8d4531" alt=“drawing” width="60%" />

## Features
There are 4PRs for the whole stack integration, this is the first one:
- [x] Enable `(grouped)AllReduce`/`Broadcast` TensorFlow Ops.
- [ ] Enable `(Grouped)Allgather`/`(Grouped)Reducescatter`/`Alltoall`/`BroadcastInplace` Tensorflow OPs.
- [ ] Add BF16 compression for `XPU` device Ops.
- [ ] Enable `(Grouped)Allreduce`/`(Grouped)Allgather`/`Broadcast`/`Alltoall`/`Reducescatter` Pytorch Ops.

**This first enabling changes include:**
- Build/installation by DPC++ compiler
`CC=icx CXX=icpx pip install --no-cache-dir intel-optimization-for-horovod`
- Register `(grouped)AllReduce`/`Broadcast` TF Ops by `XPU` device type which is an abstraction for Intel heterogeneous computation architectures and could work just like CUDA `GPU` device type.	
- Add oneCCL GPU operations which extends from Horovod GPUAllReduce and GPUBroadcast and supports prescale/postscale/batched memcpy.
<img src="https://github.com/intel-innersource/frameworks.ai.horovod/assets/20787576/5646e427-c69a-46e9-a493-1f1579c436a9" alt=“drawing” width="80%" />

- Change examples/UT to support running with `XPU` device type. The only change is `list_physical_devices` and `set_visible_devices()` by `XPU` device type.
```
    device_name = 'XPU' if hvd.sycl_built() else 'GPU'
    gpus = tf.config.experimental.list_physical_devices(device_name)
    for gpu in gpus:
        tf.config.experimental.set_memory_growth(gpu, True)
    if gpus:
        tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], device_name)
```
## Software requirement

| Software                         | Installation requirement                                     |
| -------------------------------- | ------------------------------------------------------------ |
| Intel® oneAPI Base Toolkit  | [Install Intel® oneAPI Base Toolkit](https://github.com/intel/intel-extension-for-tensorflow/blob/r1.2/docs/install/install_for_gpu.md#install-oneapi-base-toolkit-packages) |
| TensorFlow                       | [Install tensorflow >= 2.12.0](https://www.tensorflow.org/install) |
| Intel® Extension for TensorFlow* | [Install Intel® Extension for TensorFlow*](https://github.com/intel/intel-extension-for-tensorflow/tree/r1.2#install) |
| System                           | Ubuntu 22.04, RedHat 8.6 (64-bit), SUSE Linux Enterprise Server(SLES) 15 SP3/SP4 |
| GPU Driver                       | Refer to the [Installation Guides](https://dgpu-docs.intel.com/installation-guides/index.html#intel-data-center-gpu-max-series) for latest driver installation. If install the verified Intel® Data Center GPU Max Series/Intel® Data Center GPU Flex Series [602](https://dgpu-docs.intel.com/releases/stable_602_20230323.html), please append the specific version after components. |

## Examples

#### RN50 training
- The only change is `list_physical_devices` and `set_visible_devices()` by `XPU` device type. Code piece: 
```  
    device_name = 'XPU' if hvd.sycl_built() else 'GPU'
    gpus = tf.config.experimental.list_physical_devices(device_name)
    for gpu in gpus:
        tf.config.experimental.set_memory_growth(gpu, True)
    if gpus:
        tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], device_name)
```
- Full example in https://github.com/intel-innersource/frameworks.ai.horovod/blob/master/examples/tensorflow2/tensorflow2_synthetic_benchmark.py.

zhangxiaoli73 commented:
## Motivation

This PR aims to bring **Intel® oneAPI (DPC++ compiler/oneCCL/IMPI included)** software stack into Horovod to make distributed TensorFlow/PyTorch workloads run on [Intel GPU devices](https://www.intel.com/content/www/us/en/products/details/discrete-gpus/data-center-gpu.html) together with [Intel® Extension for TensorFlow*](https://github.com/intel/intel-extension-for-tensorflow) and  [Intel® Extension for PyTorch*](https://github.com/intel/intel-extension-for-pytorch), which are open-source solutions to run TensorFlow and PyTorch applications on Intel AI hardware.

**Intel® oneAPI** is an initiative to deliver a unified and open programming experience to developers on any architecture, such as CPUs, GPUs, FPGAs, and AI accelerators. Intel is promoting oneAPI as the industry standard and delivers Intel oneAPI products (an implementation of oneAPI standard) for high-performance heterogeneous computing. 

DPC++ compiler/oneCCL/IMPI packaged into oneAPI Base Tookit are core set of tools and libraries for developing high-performance computing and cross-devices communication applications. **DPC++** compiler is an extension of C++ for heterogeneous computing and based on SYCL. **OneCCL** (oneAPI Collective Communications Library) and **IMPI** provide high-performance communication patterns for distributed workloads in Intel® GPU cluster. 

To bring Intel GPU devices into TensorFlow and PyTorch community for AI workload acceleration, Intel promoted a new user visible **`XPU` device type** as a device abstraction for Intel heterogeneous computation architectures and implemented corresponding device runtime in **Intel® Extension for TensorFlow*** and **Intel® Extension for PyTorch***. 

This diagram provides a summary of the above package ecosystem. 
<img src="https://github.com/intel-innersource/frameworks.ai.horovod/assets/20787576/22bf1207-c04c-4a16-8c48-283ddf8d4531" alt=“drawing” width="60%" />

## Features
There are 4PRs for the whole stack integration, this is the first one:
- [x] Enable `(grouped)AllReduce`/`Broadcast` TensorFlow Ops.
- [ ] Enable `(Grouped)Allgather`/`(Grouped)Reducescatter`/`Alltoall`/`BroadcastInplace` Tensorflow OPs.
- [ ] Add BF16 compression for `XPU` device Ops.
- [ ] Enable `(Grouped)Allreduce`/`(Grouped)Allgather`/`Broadcast`/`Alltoall`/`Reducescatter` Pytorch Ops.

**This first enabling changes include:**
- Build/installation by DPC++ compiler
`CC=icx CXX=icpx pip install --no-cache-dir intel-optimization-for-horovod`
- Register `(grouped)AllReduce`/`Broadcast` TF Ops by `XPU` device type which is an abstraction for Intel heterogeneous computation architectures and could work just like CUDA `GPU` device type.	
- Add oneCCL GPU operations which extends from Horovod GPUAllReduce and GPUBroadcast and supports prescale/postscale/batched memcpy.
<img src="https://github.com/intel-innersource/frameworks.ai.horovod/assets/20787576/5646e427-c69a-46e9-a493-1f1579c436a9" alt=“drawing” width="80%" />

- Change examples/UT to support running with `XPU` device type. The only change is `list_physical_devices` and `set_visible_devices()` by `XPU` device type.
```
    device_name = 'XPU' if hvd.sycl_built() else 'GPU'
    gpus = tf.config.experimental.list_physical_devices(device_name)
    for gpu in gpus:
        tf.config.experimental.set_memory_growth(gpu, True)
    if gpus:
        tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], device_name)
```
## Software requirement

| Software                         | Installation requirement                                     |
| -------------------------------- | ------------------------------------------------------------ |
| Intel® oneAPI Base Toolkit  | [Install Intel® oneAPI Base Toolkit](https://github.com/intel/intel-extension-for-tensorflow/blob/r1.2/docs/install/install_for_gpu.md#install-oneapi-base-toolkit-packages) |
| TensorFlow                       | [Install tensorflow >= 2.12.0](https://www.tensorflow.org/install) |
| Intel® Extension for TensorFlow* | [Install Intel® Extension for TensorFlow*](https://github.com/intel/intel-extension-for-tensorflow/tree/r1.2#install) |
| System                           | Ubuntu 22.04, RedHat 8.6 (64-bit), SUSE Linux Enterprise Server(SLES) 15 SP3/SP4 |
| GPU Driver                       | Refer to the [Installation Guides](https://dgpu-docs.intel.com/installation-guides/index.html#intel-data-center-gpu-max-series) for latest driver installation. If install the verified Intel® Data Center GPU Max Series/Intel® Data Center GPU Flex Series [602](https://dgpu-docs.intel.com/releases/stable_602_20230323.html), please append the specific version after components. |

## Examples

#### RN50 training
- The only change is `list_physical_devices` and `set_visible_devices()` by `XPU` device type. Code piece: 
```  
    device_name = 'XPU' if hvd.sycl_built() else 'GPU'
    gpus = tf.config.experimental.list_physical_devices(device_name)
    for gpu in gpus:
        tf.config.experimental.set_memory_growth(gpu, True)
    if gpus:
        tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], device_name)
```
- Full example in https://github.com/intel-innersource/frameworks.ai.horovod/blob/master/examples/tensorflow2/tensorflow2_synthetic_benchmark.py.

maxhgerlach commented:
Thank you for breaking up the planned additions into multiple PRs! It may still take a while until somebody will be able to review these changes.

A couple of preliminary remarks:
- All commits will need to be signed off (DCO) before they can be merged to master. Please see this check for a how to: https://github.com/horovod/horovod/pull/3935/checks?check_run_id=13996532628
- At the same time you could rebase the branch to current master, which should fix the documentation build.
- There are two drawings in the description that don't show up for me, probably because they are not on public GitHub. https://github.com/intel-innersource/frameworks.ai.horovod/assets/20787576/22bf1207-c04c-4a16-8c48-283ddf8d4531 and https://github.com/intel-innersource/frameworks.ai.horovod/assets/20787576/5646e427-c69a-46e9-a493-1f1579c436a9
- Ultimately, it would be great if some tests running on Intel GPUs could be added to Horovod's CI. Would it be possible for Intel to provide such an integration?
- As a first step it might make sense to check build stability on CPU (with a variation of `Dockerfile.test.gpu`) without actually running tests, which could be done directly in GitHub CI.

zhangxiaoli73 commented:
@maxhgerlach Thanks for your quick response and remarks.

- It's a real good reminder to us for `DCO` and documentation build. We will check and fix.
- I have updated drawings link and hope work for you now, please help check.
- For Horovod CI, we can add build and some tests for Intel GPUs. However, as Horovod doesn't have any Intel GPUs which means Intel GPUs tests cannot work in Horovod CI, so would you still need us to add tests and validate those tests in future like when you have Intel GPUs?

maxhgerlach commented:
Thanks for the Dockerfile integration, @zhangxiaoli73! I started looking over the changes and will continue to do so, but it will take a while.

EnricoMi commented on docs/keras.rst:
```diff
@@ -25,11 +25,14 @@ Horovod supports Keras and regular TensorFlow in similar ways. To use Horovod wi
    .. code-block:: python
 
-       gpus = tf.config.experimental.list_physical_devices('GPU')
+       device_type = 'XPU' if hvd.sycl_built() else 'GPU'
```
From a user perspective, `sycl_built` seems very opaque. Can this method name be a bit more meaningful in the sense why am I calling it? How does `sycl` relate to the abstraction layer that provides `XPU`s?

EnricoMi replied:
This example has the drawback that users that want to use `XPU`s do not realize when `hvd.sycl_built` returns false and the example silently falls back to GPU. With `device_type = 'XPU'` they will get an error in that situation and realize Horovod is not build with `sycl` support.

I'd rather prefer the example to look like
```suggestion
       device_type = 'GPU'  # use 'XPU' here if you are using Intel GPU devices with oneAPI
```

zhangxiaoli73 replied:
@EnricoMi Agreed, we have changed this example docs to your suggestions. Common users only need care about "GPU" and "XPU". "sycl_built" should be transparent to users. Thanks for your review and we appreciate your any comments.

EnricoMi replied:
Looks better, thanks!

EnricoMi reviewed (commented):
Some minor comments on examples and usability.

EnricoMi commented on test/parallel/test_tensorflow.py:
```diff
@@ -53,9 +53,11 @@ class TensorFlowTests(BaseTensorFlowTests):
 
     def test_gpu_required(self):
-        if not tf.test.is_gpu_available(cuda_only=True):
+        if not tf.test.is_gpu_available(cuda_only=True) and not hvd.sycl_built():
```
So tensorflow thinks, there is no GPU available, but there is when sycl is built?

LuFinch commented:
@EnricoMi @maxhgerlach  
Hi guys, we decide to cancel upstream our code because we have added a lot of code changes in `horovod/tensorFlow/mpi_ops.cc` to support our XPU as [PJRT](https://opensource.googleblog.com/2023/06/accelerate-jax-models-on-intel-gpus-via-pjrt.html) [NextPluggableDevice Mechanism](https://docs.google.com/document/d/1S1pdXUUNYYr5cpf9F2ovGoP-Gnpxb-698tdDqaHZTok/edit?resourcekey=0-sqVzq9GnBLREq1tkzezzIQ#heading=h.dw9zn5llfpgj) and fit OpenXLA in TensorFlow 2.15. Here we added a lot of TensorFlow C API changes in horovod which breaks the original logic and might be hard to review/upstream. 

We are sorry for wasting your time and great thanks for your response and effort on review/comments!  You could close this JIRA!
